storage/src/mmr/bitmap.rs

```
@@ -0,0 +1,315 @@
+    pub fn bit_count(&self) -> u64 {
+        (self.bitmap.len() * 8 - Self::CHUNK_SIZE * 8 + self.next_bit) as u64
+    }
+
+    /// Return the last chunk of the bitmap as a slice.
+    fn last_chunk(&self) -> &[u8] {
+        let len = self.bitmap.len();
+        &self.bitmap[len - Self::CHUNK_SIZE..len]
+    }
+
+    /// Return the last chunk of the bitmap as a mutable slice.
+    fn last_chunk_mut(&mut self) -> &mut [u8] {
+        let len = self.bitmap.len();
+        &mut self.bitmap[len - Self::CHUNK_SIZE..len]
+    }
+
+    /// Commit the last chunk of the bitmap to the Merkle tree and initialize the next chunk.
+    fn commit_last_chunk(&mut self, hasher: &mut H) {
+        let chunk = H::Digest::try_from(self.last_chunk()).unwrap();
+        self.mmr.add(hasher, &chunk);
+        self.next_bit = 0;
+        self.bitmap.extend(vec![0u8; Self::CHUNK_SIZE]);
+    }
+
+    /// Efficiently add a digest-sized chunk of bits to the bitmap. Assumes we are at a chunk
+    /// boundary (that is, `self.next_bit` is 0) and panics otherwise.
+    pub fn append_chunk_unchecked(&mut self, hasher: &mut H, chunk: &H::Digest) {
+        assert!(
+            self.next_bit == 0,
+            "cannot add chunk when not chunk aligned"
+        );
+
+        self.last_chunk_mut().copy_from_slice(chunk.as_ref());
+        self.commit_last_chunk(hasher);
+    }
+
+    /// Efficiently add a byte's worth of bits to the bitmap. Assumes self.next_bit is currently
+    /// byte aligned, and panics otherwise.
+    pub fn append_byte_unchecked(&mut self, hasher: &mut H, byte: u8) {
+        assert!(
+            self.next_bit % 8 == 0,
+            "cannot add byte when not byte aligned"
+        );
+
+        let chunk_byte = self.next_bit / 8;
+        self.last_chunk_mut()[chunk_byte] = byte;
+        self.next_bit += 8;
+        if self.next_bit == Self::CHUNK_SIZE * 8 {
+            self.commit_last_chunk(hasher);
+        }
```

I wonder if we should `panic!` if `next_bit` is ever larger than `CHUNK_SIZE * 8`.

Might not be worth checking but maybe it is

---

I'm generally a fan of asserting on invariants and failing fast when violated, so added a couple assertions around this.

---

My thinking was maybe leave it here for now and consider a broader refactor once the mutable store is functional.